### ISSUE TITLE
Supporting request matchers in DSL

### ISSUE THREAD

**Issue by tommysitu** [SpectoLabs/hoverfly-java#102]
Starting from v0.5.0, the simulation scheme allows matchers in the request data, we should make use of it from the DSL. 

There are a few changes I am proposing: 
1. Matcher for url: 
```
service("www.test.com") // match on any http protocol
service("http://www.test.com") // match exactly on scheme=http, and destination=www.test.com
service(any()) // match on any scheme and any destination

service("www.test.co*")  // enable glob matcher, that will match any destination begins with 'www.test.co' 

// or maybe a bit Java-ish: 
service(startsWith("www.test.co"))  // same as "www.test.co*"
service(endsWith("test.com")) // same as "*test.com"
service(contains("test")) // same as  "*test*"
service(matches("www.*-test.com")) // explicitly set glob matching

// then later on when we support regex matcher
service(matchesRegex("www.[test|fail].com")) 
```
2. Matcher for HTTP method: 
```
service("www.test.com")
 .anyMethod("/api/foo")   // match any HTTP methods made to '/api/foo' endpoint
```

3. Matcher for url path:
```
 .get("/api/v1/foo") // match '/api/v1/foo' exactly 

 .get(any()) // match GET request to any url path

 .get("/api/*/foo") // match paths such as '/api/v1/foo', '/api/v2/foo', behind the scene, it uses glob matcher

// or 
 .get(matches("/api/*/foo")) 
```

4. Matcher for query params:

```
.get("/api/v1/foo")
.queryParam(any()) // match request with any params

.queryParam(any(), "foo")  // match any query has value "foo"

.queryParam("foo", any()) // match any query has key "foo" with any value
```

5. Matcher for the body:
```
// JSON 
.body(matchesJsonPath("<jsonpath expression>"))
.body(equalsToJson("<request body in JSON>"))
.body(equalsToJson(obj)) // will serialize the object to JSON 

// XML 
.body(matchesXmlPath("<xmlpath expression>"))
.body(equalsToXml("<request body in XML>"))
.body(equalsToXml(obj)) // will serialize the object to XML 

// Others
 .body(any())  // match request with any body, this will include empty or null body
```

6. Maybe we can have a builder for JSONPath expression?
```
.body(matchesJson().arrayNode("items").get(0).getKey("price").equalsTo(100));
```
maybe it is too much.

7. If `.body()` or `.queryParams()` is absent, we are assuming that the user want to match request that has no body or query params
  
8. Glob matcher can be built using these methods:
 - `any()` -> `*`
 - `startsWith("foo")` -> `foo*`
 - `endsWith("foo")` -> `*foo`
 - `contains("foo")` -> `*foo*`
 - `matches("fo*o")` -> `fo*o`

9. I don't think regex matcher can be fully supported, as the simulation JSON takes Golang regex instead of Java regex. Maybe we can write a converter.

**Comment by benjih** [comment]
Hey @tommysitu,

This is a nice proposal. Regarding number`7`, Hoverfly works opposite to this. The absence of these fields means they are no checked. If you wanted to match a request with an empty body, you would need to specify the body, rather than leaving it out.

**Comment by tommysitu** [comment]
@benjih do you need to set `body: ''` to match on an empty body? If that's the case, 
I would change item 7 to: 
```
.body(empty())
.queryParams(empty())
```

**Comment by mogronalol** [comment]
I think this looks good, but I don't think `globMatcher` should be inferred. I think behind the scenes we should have something like:

```
requestMatchers.globMatcher("/api/*")
requestMatchers.exactMatcher("/api")
```

Then in the DSL, the argument can only take:

```
public void get(final Matcher matcher)
public void get(final String string) // Creates a exactMatcher internally
```

We also need to make sure to limit what's possible for each part of a request. So making it impossible to pass a `jsonPath` matcher into the `path` part of a request.

**Comment by tommysitu** [comment]
Released as v0.6.0
